garak/generators/huggingface.py

```diff
@@ -50,53 +52,134 @@ def _set_hf_context_len(self, config):
             if isinstance(config.n_ctx, int):
                 self.context_len = config.n_ctx
 
+    def _gather_hf_params(self, hf_constructor: Callable):
+        # this may be a bit too naive as it will pass any parameter valid for the pipeline signature
+        # this falls over when passed `from_pretrained` methods as the callable model params are not explicit
+        params = self.hf_args
+        if params["device"] is None:
+            params["device"] = self.device
+
+        args = {}
+
+        parameters = inspect.signature(hf_constructor).parameters
+
+        if "model" in parameters:
+            args["model"] = self.name
+            # expand for
+            parameters = {"do_sample": True} | parameters
+        else:
+            # callable is for a Pretrained class also map standard `pipeline` params
+            from transformers import pipeline
+
+            parameters = (
+                {"low_cpu_mem_usage": True}
+                | parameters
+                | inspect.signature(pipeline).parameters
+            )
+
+        for k in parameters:
+            if k == "model":
+                continue  # special case `model` comes from `name` in the generator
+            if k in params:
+                val = params[k]
+                if k == "torch_dtype" and hasattr(torch, val):
+                    args[k] = getattr(
+                        torch, val
+                    )  # some model type specific classes do not yet support direct string representation
+                    continue
+                if (
+                    k == "device"
+                    and "device_map" in parameters
+                    and "device_map" in params
+                ):
+                    # per transformers convention hold `device_map` before `device`
+                    continue
+                args[k] = params[k]
+
+        return args
+
+    def _select_hf_device(self):
+        """Determine the most efficient device for tensor load, hold any existing `device` already selected"""
+        import torch.cuda
+
+        selected_device = None
+        if self.hf_args["device"] is not None:
+            if isinstance(self.hf_args["device"], int):
```

Comment:
Should check >= 0. Sometimes, folks use `-1` as a stand-in for CPU (which is not a good practice, but still exists out there.)

Reply:
`-1` will now raise an exception is offered in a config file:
```
$ cat garak/garak.site.yaml
plugins:
  generators:
    huggingface:
      hf_args:
        torch_dtype: float16
      Model:
        hf_args:
            device: -1
```
Resulting in:
```
$ ./garak_debug.py -m huggingface.Model -n meta-llama/Llama-2-7b-chat-hf -g 1 -p continuation
garak LLM vulnerability scanner v0.9.0.13.post2 ( https://github.com/leondz/garak ) at 2024-06-18T10:06:06.476749
🦜 loading generator: Hugging Face 🤗 model: meta-llama/Llama-2-7b-chat-hf
Invalid device string: 'cuda:-1'
```

Reply:
Let's throw out a clear error if they send `-1`. E.g.
```suggestion
            if isinstance(self.hf_args["device"], int):
              if self.hf_args["device"] < 0:
                msg = f"device {self.hf_args["device"]} requested but CUDA device numbering starts at zero. Use 'device: cpu' to request CPU."
                logging.critical(msg)
                raise ValueError(msg)
```